### ISSUE TITLE
[strict-boolean-expressions] Breaks type when using allowComparingNullableBooleansToTrue

### ISSUE THREAD

**Issue by paleite** [typescript-eslint/typescript-eslint#3783]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JS
{
  "rules": {
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": [
      "error",
      {
        allowComparingNullableBooleansToTrue: false,
        allowComparingNullableBooleansToFalse: false,
      },
    ],
  }
}
```

```TS
const isTrue = (x: boolean | undefined): boolean => x === true;
```

**Expected Result**

I would've expected either one of the following fixes to be output

```TS
// fall back to false, since undefined is falsey
const isTrue = (x: boolean | undefined): boolean => x ?? false;
// coerce to boolean
const isTrue = (x: boolean | undefined): boolean => !!x;
// convert to boolean
const isTrue = (x: boolean | undefined): boolean => Boolean(x);
// leave as-is
const isTrue = (x: boolean | undefined): boolean => x === true;
```

**Actual Result**

```TS
const isTrue = (x: boolean | undefined): boolean => x;
```

**Additional Info**

The TypeScript compiler now gives an error:

```
Type 'boolean | undefined' is not assignable to type 'boolean'.
  Type 'undefined' is not assignable to type 'boolean'. ts(2322)
```

**Versions**

| package                            | version   |
| ---------------------------------- | --------- |
| `@typescript-eslint/eslint-plugin` | `4.29.2`  |
| `@typescript-eslint/parser`        | `4.29.2`  |
| `TypeScript`                       | `4.3.5`   |
| `ESLint`                           | `7.30.0`  |
| `node`                             | `14.16.1` |

**Comment by bradzacher** [comment]
The rule does not consider the context of the expression.
Some cases would be easy to handle:
```ts
declare const bool: boolean | undefined;

// cases where it's easy to tell it's required, and thus can fix differently or not report
(): boolean => bool === true;
function f(): boolean { return bool === true }

declare function f2(arg: boolean): void;
f2(bool === true);

declare function Component(props: {bool: boolean}): JSX.Element;
<Component bool={bool === true} />

const x1: boolean = bool === true;
const x2: {bool: boolean} = { bool: bool === true };


// cases where it's easy to tell it's not required and thus fix is safe
if (bool === true) {}
(bool === true) ? null : null;
// all of the above except the annotation is `boolean | undefined` instead of `boolean`
```

But some cases would be hard-to-impossible to handle (really any case with a layer of indirection:
```ts
declare const bool: boolean | undefined;

function f(): boolean {
  const result = bool === true;
  return result;
}
```

We could add some handling to convert the automatic fixer to a suggestion fixer if we think it's not safe.
Or we could just make the fixer always a suggestion fixer.

The former would probably make for a better UX, but it'll be a lot more work.

**Comment by paleite** [comment]
Amazing! Thanks for looking into this so thoroughly 🙂 

Personally, I'd be more than satisfied if, as a first iteration in fixing this, the autofixer was disabled for this case.

Thanks again for doing such a great job! 💯

**Comment by phaux** [comment]
wrong rule in issue title btw

**Comment by Vinccool96** [comment]
Wait, hold on. In the example provided, the rule explicitly disallows comparing nullable booleans to literals. Isn't this working exactly as intended? Or do we want a suggestion to fallback to `false` when `allowComparingNullableBooleansToX` is enabled?

**Comment by paleite** [comment]
@Vinccool96 : You're right that the rule is identifying exactly what it's supposed to. The issue I reported isn't about the diagnostic itself, but about the autofix that gets applied, which changes the type of the effectively returned value.

In my repro, the autofix changes the effectively returned value's type from `boolean` to `boolean | undefined`:

```ts
const isTrue = (x: boolean | undefined): boolean => x === true;
// autofixes to:
const isTrue = (x: boolean | undefined): boolean => x;
```

So the diagnostic is correct, but the autofix is unsafe here because it introduces a compiler error. The fix needs to preserve the boolean result, e.g. `x ?? false`, `!!x`, `Boolean(x)`, or be suggestion-only.
